src/cmd/serve.rs

```diff
@@ -503,82 +587,63 @@ pub fn serve(
-                    eprintln!("Failed to open URL in your browser: {err}");
-                }
-
-                axum::serve(listener, app).await.expect("Could not start web server");
-            });
-        });
-
-        // The websocket for livereload
-        let ws_server = WebSocket::new(|output: Sender| {
-            move |msg: Message| {
-                if msg.into_text().unwrap().contains("\"hello\"") {
-                    return output.send(Message::text(
-                        r#"
-                        {
-                            "command": "hello",
-                            "protocols": [ "http://livereload.com/protocols/official-7" ],
-                            "serverName": "Zola"
-                        }
-                    "#,
-                    ));
-                }
-                Ok(())
+    // Create broadcast channel for WebSocket live reload
+    let (reload_tx, _) = broadcast::channel::<String>(100);
+    let broadcaster = reload_tx.clone();
+
+    // Shutdown signal for graceful termination
+    let shutdown_flag = Arc::new(AtomicBool::new(false));
+    let shutdown_flag_server = shutdown_flag.clone();
+
+    // Start Axum server in a separate thread
+    thread::spawn(move || {
+        let rt = tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .expect("Could not build tokio runtime");
+
+        rt.block_on(async {
+            let state = Arc::new(AppState { static_root, base_path, reload_tx });
+
+            let app = Router::new()
+                .route("/livereload.js", get(serve_livereload_js))
+                .route("/livereload", get(ws_handler))
+                .fallback(handle_request)
+                .layer(middleware::map_response(error_injection_middleware))
+                .with_state(state);
+
+            let listener = tokio::net::TcpListener::bind(&bind_address)
+                .await
+                .expect("Could not bind to address");
+
+            let local_addr = listener.local_addr().unwrap();
+
+            println!(
+                "Web server is available at {} (bound to {})\n",
+                &constructed_base_url.replace(&bind_address.to_string(), &local_addr.to_string()),
+                &local_addr
+            );
+            if open && let Err(err) = open::that(&constructed_base_url) {
+                eprintln!("Failed to open URL in your browser: {err}");
             }
-        })
-        .unwrap();
 
-        let broadcaster = ws_server.broadcaster();
-
-        let ws_server = ws_server
-            .bind(&*ws_address)
-            .map_err(|_| anyhow!("Cannot bind to address {} for the websocket server. Maybe the port is already in use?", &ws_address))?;
-
-        thread::spawn(move || {
-            ws_server.run().unwrap();
+            axum::serve(listener, app)
+                .with_graceful_shutdown(async move {
```

Comment:
why do we need graceful shutdown for the dev server

Reply:
If you'd rather have errors in the console, every time you restart the dev server because the livereload socket was pulled out from under -- sure, but it's an easy fix to do it right.